src/mlir/consumer.h

```diff
@@ -0,0 +1,149 @@
+// Copyright Microsoft and Project Verona Contributors.
+// SPDX-License-Identifier: MIT
+
+#pragma once
+
+#include "generator.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/MLIRContext.h"
+#include "parser/ast.h"
+
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/Error.h"
+
+#include <string>
+
+namespace mlir::verona
+{
+  /**
+   * Field offset maps between field names, types and their relative position.
+   *
+   * FIXME: This will need to be better encapsulated, but we'll expand it when
+   * we start implementing classes fully, and dynamic dispatch.
+   */
+  struct FieldOffset
+  {
+    llvm::SmallVector<llvm::StringRef, 1> fields;
+    llvm::SmallVector<Type, 1> types;
+  };
+
+  /**
+   * AST Consumer
+   */
+  class ASTConsumer
+  {
+    ASTConsumer(MLIRContext* context) : gen(context) {}
+
+    /// MLIR Generator
+    MLIRGenerator gen;
+
+    /// Map for each type which fields does it have.
+    std::map<llvm::StringRef, FieldOffset> classFields;
```

Comment:
Why `map` and not `unordered_map`, do we care about ordering here?

`llvm::StringRef` is quite dangerous to use as a map key, because it is a non-owning value and so there's an undocumented invariant here that the lifetime of every class name (wherever it comes from) must outlive the class.  I guess that's okay because it comes from the AST and this class is a transient thing that consumes an AST?

If you want to key on strings (which I'd prefer you didn't and instead keyed on interned string IDs), then `llvm::StringMap` is probably what you actually want.